### PR TITLE
Fix: Allow accelerators as list-of-dicts in YAML; add parsing, schema, and tests (#7098)

### LINKED ISSUE #7098: [core] Resources specified as a dict work, but not as a list
Specifying multiple candidate resources non-inline in YAML is a bit inconsistent:

```
# Works - parsed as dict of dicts; applies any-of behavior.
resources:
  accelerators:
    T4: 1
    L4: 1

# Does not work - list of dicts does not work: sky.exceptions.InvalidSkyPilotConfigError: Invalid resources YAML: [{'T4': 1}, {'L4': 1}] is not valid under any of the given schemas. Check problematic field(s): $.accelerators
resources:
  accelerators:
    - T4: 1
    - L4: 1

# Works - gets parsed as list of str; applies ordered behavior.
resources:
  accelerators:
    - T4:1
    - L4:1
```

We should probably allow the list of dicts syntax as well so it's consistent.

### PR DESCRIPTION
**Pull request by pravarpathania**
## **Summary**

This PR adds support for specifying accelerators in YAML as a list of dicts, enabling more intuitive accelerator configuration formats while maintaining full backward compatibility.

## **Problem**

The YAML schema rejected accelerator specifications given as a list of dicts (e.g., `accelerators: [{"T4": 1}, {"L4": 1}]`), which was inconsistent with other supported formats (dict-of-dicts and list-of-strings). Users requested this format for better ergonomics when specifying multiple accelerators.

### **Schema Enhancement (`sky/utils/schemas.py`)**

* Modified `get_resources_schema()` to accept array items that can be either strings or objects  
* Updated schema to allow both integer and float values for fractional GPU support  
* Made schema more permissive to enable custom validation with better error messages

### **Parser Enhancement (`sky/resources.py`)**

* Extended `Resources.from_yaml_config()` to handle dict items in accelerator lists  
* Added comprehensive validation in `_validate_accelerators_config()` with proper error messages  
* Implemented unified parsing logic to reduce code duplication  
* Added support for fractional accelerator values (0.5, 1.5, etc.)  
* Ensured consistent return types (always returns `set` for multiple accelerators)

### **Validation Improvements**

* Custom validation now runs before schema validation for better error messages  
* Consistent `AssertionError` style error handling throughout  
* Proper validation of null values, empty dicts, and invalid types  
* Support for both integer and float accelerator counts

### **Comprehensive Testing**

* Added unit tests for all new formats and edge cases  
* Added integration tests via `test_optimizer_dryruns.py`  
* Added validation error tests to ensure proper error handling  
* Tests cover fractional GPU support

## **Files Modified**

* `sky/utils/schemas.py` — enhanced accelerator schema support  
* `sky/resources.py` — added list-of-dicts parsing and validation  
* `tests/unit_tests/test_resources.py` — comprehensive test coverage  
* `tests/test_optimizer_dryruns.py` — integration test updates

## **Supported Formats**

All of the following formats are now supported:

\# Original formats (unchanged)  
accelerators: "T4:1"  
accelerators: { T4: 1, L4: 2 }  
accelerators: \["T4:1", "L4:2"\]

\# New list-of-dicts format  
accelerators:  
  \- T4: 1  
  \- L4: 2

\# Mixed format support  
accelerators:  
  \- T4: 1  
  \- "V100:2"  
  \- A100: 1

\# Fractional GPU support  
accelerators:  
  \- V100: 0.5  
  \- T4: 1.5

\# Error handling for invalid values  
accelerators:  
  \- T4: 1  
  \- V100: null  \# Properly rejected with clear error message

### **Unit Tests**

pytest tests/unit\_tests/test\_resources.py::test\_accelerators\_list\_of\_dicts  
pytest tests/unit\_tests/test\_resources.py::test\_accelerators\_mixed\_list\_format  
pytest tests/unit\_tests/test\_resources.py::test\_accelerators\_fractional\_support  
pytest tests/unit\_tests/test\_resources.py::test\_accelerators\_type\_validation  
pytest tests/unit\_tests/test\_resources.py::test\_accelerators\_equivalence

### **Integration Tests**

pytest tests/test\_optimizer\_dryruns.py::test\_parse\_accelerators\_list\_of\_dicts\_from\_yaml  
pytest tests/test\_optimizer\_dryruns.py::test\_parse\_accelerators\_fractional\_from\_yaml  
pytest tests/test\_optimizer\_dryruns.py::test\_parse\_accelerators\_validation\_errors\_from\_yaml

### **Regression Tests**

pytest tests/unit\_tests/test\_resources.py \-k "accelerator"  
pytest tests/test\_optimizer\_dryruns.py \-k "accelerator"

### **Enhanced Error Messages**

* Clear, actionable error messages for validation failures  
* Consistent error handling across all accelerator formats  
* Proper null value detection and reporting

### **Fractional GPU Support**

* Support for fractional accelerator values (0.5, 1.5, etc.)  
* Validation ensures positive numbers (not just integers)  
* Consistent behavior across all specification formats

### **Format Equivalence**

* All supported formats produce identical results  
* Seamless conversion between different input formats  
* Unified internal representation

## **Backward Compatibility**

This change is completely backward compatible. All existing accelerator specification formats continue to work unchanged. The implementation is purely additive.

## Testing Checklist

- [ ] Code formatting: install pre-commit (auto-check on commit) or `bash format.sh`
- [ ] Any manual or new tests for this PR (specified below)
- [ ] All smoke tests: `/smoke-test` (CI) or `pytest tests/test_smoke.py` (local)
- [ ] Relevant individual tests: `/smoke-test -k test_name` (CI) or `pytest tests/test_smoke.py::test_name` (local) 
- [ ] Backward compatibility: `/quicktest-core` (CI) or `pytest tests/smoke_tests/test_backward_compat.py` (local)

## **AWS Test Failures (Note for Maintainers)**

Some integration tests in `tests/test_optimizer_dryruns.py` currently fail due to missing AWS credentials:
botocore.exceptions.NoCredentialsError: Unable to locate credentials
sky.exceptions.CloudError: botocore error (NoCredentialsError): Unable to locate credentials

This is **not a code bug** but an infrastructure issue. Maintainers should either:

- Provide valid AWS credentials in the CI environment, or 
- Mock/skip tests requiring AWS in dry-run scenarios.

## **Fixes**

- Fixes #7098

### CODE REVIEW

**Comment by Michaelvll** [comment]
Thanks for submittin the PR @pravarpathania! There seems to be a large amount of changes in this PR. Could you help limit it to the necessary changes only? : )

**Comment by pravarpathania** [comment]
Yes, 

> Thanks for submittin the PR @pravarpathania! There seems to be a large amount of changes in this PR. Could you help limit it to the necessary changes only? : )

Thanks for the review @Michaelvll! You're right, I noticed that this PR included a lot of unrelated changes (most likely from formatting and history divergence).

I'll clean up the branch so that it only contains the schema/parser changes and the new tests, then force-push an updated commit.

**Comment by pravarpathania** [comment]
> Thanks for submittin the PR @pravarpathania! There seems to be a large amount of changes in this PR. Could you help limit it to the necessary changes only? : )

@Michaelvll, I have rebased and removed unrelated formatting/history changes, so this PR now only contains the schema/parser updates, new tests, and the small mock.patch fix in test_resources.py to make AWS deploy tests hermetic.

Would you mind taking another look when you get a chance?

File: sky/resources.py [hunk 1]
```diff
@@ -2044,12 +2044,28 @@ def _override_resources(
...
                     accelerators, set):
                 accelerators_list = []
-                for accel_name in accelerators:
-                    parsed_accels = cls._parse_accelerators_from_str(accel_name)
-                    accelerators_list.extend(parsed_accels)
-            else:
-                assert False, ('Invalid accelerators type:'
-                               f'{type(accelerators)}')
+                for accel_item in accelerators:
+                    if isinstance(accel_item, str):
```
Comment (kyuds) [inline sky/resources.py]: I see a bit of duplicated logic here, could we try and think of a way to convert a `dict` `accel_item` to string, and then call `cls._parse_accelerators_from_str` and `accelerators_list.extend` once?

File: sky/resources.py [hunk 2]
```diff
@@ -2044,12 +2044,28 @@ def _override_resources(
...
+                        for acc_name, acc_count in accel_item.items():
+                            if acc_count is None:
+                                acc_str = acc_name
+                            else:
+                                acc_str = f'{acc_name}:{acc_count}'
+                            parsed_accels = cls._parse_accelerators_from_str(
+                                acc_str)
+                            accelerators_list.extend(parsed_accels)
+                    else:
+                        with ux_utils.print_exception_no_traceback():
```
Comment (kyuds) [inline sky/resources.py]: we throw AssertionErrors using `assert False, {error message}` when when we are checking for accelerator types (see the assertion statement for the outer for loop). Could we follow similar style and change this to an assertion error? We could have the error message like: `Invalid accelerators item type: <type>. Expected str or dict.`

File: sky/utils/schemas.py [hunk 3]
```diff
@@ -409,7 +409,19 @@ def get_resources_schema():
...
                     'items': {
-                        'type': 'string',
+                        'anyOf': [{
+                            'type': 'string',
+                        }, {
+                            'type': 'object',
+                            'required': [],
+                            'additionalProperties': {
+                                'anyOf': [{
+                                    'type': 'null',
```
Comment (kyuds) [inline sky/utils/schemas.py]: why type is null?

File: sky/resources.py (hunk 1 above)
Comment (pravarpathania) [reply sky/resources.py]: Thanks for pointing this out! I've refactored the parsing so that dict items are first converted to strings, and then `_parse_accelerators_from_str` is called once for each item. This eliminates duplication and keeps the logic unified.

File: sky/resources.py (hunk 2 above)
Comment (pravarpathania) [reply sky/resources.py]: Updated the code to raise `AssertionError` with a clear message:  
"Invalid accelerators item type: <type>. Expected str or dict."  
This is now consistent with other validation checks in the file.

**Comment by pravarpathania** [comment]
AWS Integration Test Failures (Note for Maintainers)
Some integration tests in `tests/test_optimizer_dryruns.py` currently fail in CI due to missing AWS credentials:
